### PR TITLE
Flag for disabling random monster spawning

### PR DESCRIPTION
**Pull request by samvelyan**
Adding the `spawn_monsters` flag to NLE to be able to disable random monster spawning that aren't described in the level's des-file (see [this](https://nethackwiki.com/wiki/Monster_creation) for more info). The same flag is also added to the NetHack class (True by default in both places). Setting this flag off triggers the `set_spawn_monsters` function in `win/rl/pynethack.cc`, which updates the corresponding value in `nle_init_settings` (previously called `nle_seed_init`). This later updates an external variable declared in `nle.h` that is checked in `src/allmain.c` before random monster spawning logic.

All of this functionality is available if `NLE_ALLOW_CONTROL` macro is defined. I modified the name of `NLE_ALLOW_SEEDING` since now it controls more than just seeding.

The above-mentioned `NLE_ALLOW_CONTROL` macro is manually defined in this branch in case someone want to try it out. I'll comment it out in the final version of this PR.

### CODE REVIEW

File: src/allmain.c [hunk 1]
```diff
@@ -6,6 +6,7 @@
 /* various code that was replicated in *main.c */
 
 #include "hack.h"
+#include "nle.h"
```
Comment (heiner) [inline src/allmain.c:9]: Let's not include nle.h here.

Instead, let's move the `extern int nle_spawn_monsters;` line from nle.h into this file (with a comment).

File: src/allmain.c [hunk 2]
```diff
@@ -119,13 +120,16 @@ boolean resuming;
...
 
                     /* occasionally add another monster; since this takes
-                       place after movement has been allotted, the new
-                       monster effectively loses its first turn */
-                    if (!rn2(u.uevent.udemigod ? 25
+                       place after movement has been allotted (unless turned
+                       off in nle), the new monster effectively loses its first
+                       turn */
+                    if (nle_spawn_monsters) {
+                        if (!rn2(u.uevent.udemigod ? 25
```
Comment (heiner) [inline src/allmain.c:127]: I would suggest merging these ifs and doing `if (nle_spawn_monsters && !rn2(u.uevent.udemigod ? 25` ...

File: win/rl/pynethack.cc [hunk 3]
```diff
@@ -177,14 +177,25 @@ class Nethack
...
-#ifdef NLE_ALLOW_SEEDING
-        seed_init_.seeds[0] = core;
-        seed_init_.seeds[1] = disp;
-        seed_init_.reseed = reseed;
-        use_seed_init = true;
+#ifdef NLE_ALLOW_CONTROL
+        init_settings_.seeds[0] = core;
+        init_settings_.seeds[1] = disp;
+        init_settings_.reseed = reseed;
+        use_init_settings = true;
```
Comment (heiner) [inline win/rl/pynethack.cc:198]: Ah, this is a bit dangerous I think. `use_seed_init` used to change the seed arguments form the default. We might have to change this a bit.

File: win/rl/pynethack.cc [hunk 4]
```diff
@@ -177,14 +177,25 @@ class Nethack
         }
     }
 
+    void
+    set_spawn_monsters(bool spawn_monsters)
+    {
+#ifdef NLE_ALLOW_CONTROL
+        init_settings_.spawn_monsters = spawn_monsters;
+        use_init_settings = true;
```
Comment (heiner) [inline win/rl/pynethack.cc:185]: I think this has the side effect of changing the way seeding works, which isn't necessarily what we want here.

File: nle/nethack/nethack.py [hunk 5]
```diff
@@ -168,6 +170,8 @@ def reset(self, new_ttyrec=None, wizkit_items=None):
             _set_env_vars(self._options, self._vardir, wizkit=WIZKIT_FNAME)
         else:
             _set_env_vars(self._options, self._vardir)
+        if not self._spawn_monsters:
+            self._pynethack.set_spawn_monsters(False)
```
Comment (heiner) [inline nle/nethack/nethack.py:174]: I think it's a better idea to make it such that the underlying `_pynethack` object is either started with spawning on or off, without the ability to change this during its lifetime.

**Comment by samvelyan** [comment]
Just realised this messes up the seeding values. Specifically, when turning off monster spawning and not providing seed values, the corresponding fields in the `nle_init_settings` struct are initialised by defaults and the episode look exactly the same after every reset. This needs to be fixed.

**Review by heiner — CHANGES_REQUESTED** [review]
While this generally looks good, I think my idea of merging this setting with the seeding setting creates more complexity than I anticipated. I think that is also the reason the tests fail: Not giving a nle_seeds_init vs giving an empty/default one is a big difference, in that the empty one initializes with a seed of 0.

To keep the previous behaviour we should probably go another route: Instead of adding the spawn setting to the existing struct, it seem we will have to create two structs and add an additional argument to `nle_start`.

I would also suggest fixing the spawn value at construction of the C++ PyNetHack object -- I don't see a big benefit from being able to change spawning on resets. Let me know if you see this differently? Not allowing it would make the logic on the Python side simpler, which I think is a decent benefit. (For seeding, we can't do that unfortunately.)

**Review by samvelyan — COMMENTED** [review]
Applied all change requests.

File: include/nleobs.h [hunk 6]
```diff
@@ -45,6 +45,8 @@ typedef struct {
 #ifdef NLE_ALLOW_SEEDING
     unsigned long seeds[2]; /* core, disp */
     char reseed; /* boolean: use NetHack's anti-TAS reseed mechanism? */
+    int spawn_monsters; /* boolean: spawn monsters randomly after every step
```
Comment (heiner) [inline include/nleobs.h:48]: Did you mean for this to be in the #ifdef?

**Review by heiner — COMMENTED** [review]
Looks good, two small comments below.

File: include/nleobs.h (hunk 6 above)
Comment (samvelyan) [reply include/nleobs.h:48]: Good catch. Forgot to remove this from the struct.

**Review by samvelyan — COMMENTED** [review]
Applied both requests.

**Review by heiner — APPROVED** [review]
Looks good, thanks.
